modules/programs/neovim.nix

```diff
@@ -164,24 +156,56 @@ in {
       };
 
       extraPython3Packages = mkOption {
-        type = with types; either extraPython3PackageType (listOf package);
-        default = (_: [ ]);
+        # In case we get a plain list, we need to turn it into a function,
+        # as expected by the function in nixpkgs.
+        # The only way to do so is to call `const`, which will ignore its input.
+        type = with types;
+          let fromType = listOf package;
+          in coercedTo fromType
+          (convertListToFunctionWithWarning "extraPython3Packages")
+          (functionTo fromType);
+        default = _: [ ];
         defaultText = literalExpression "ps: [ ]";
-        example = literalExpression "(ps: with ps; [ python-language-server ])";
+        example = literalExpression
+          "python3Packages: with python3Packages; [ python-language-server ]";
         description = ''
-          A function in python.withPackages format, which returns a
-          list of Python 3 packages required for your plugins to work.
+          The extra Python 3 packages required for your plugins to work.
+          This option accepts a function that takes a Python 3 package set as an argument,
+          and selects the required Python 3 packages from this package set.
+          See the example for more info.
+
+          For backwards compatibility, this option also accepts a plain list of
+          packages, but this functionality is deprecated and might be removed in
+          a future version, because doing so might cause problems when the
+          specified packages are not compatible with the Python 3 version used
+          by Neovim.
         '';
       };
 
+      # We get the Lua package from the final package and use its
+      # Lua packageset to evaluate the function that this option was set to.
+      # This ensures that we always use the same Lua version as the Neovim package.
       extraLuaPackages = mkOption {
-        type = with types; either extraLua51PackageType (listOf package);
-        default = [ ];
-        defaultText = literalExpression "[ ]";
-        example = literalExpression "(ps: with ps; [ luautf8 ])";
+        type = with types;
+          let fromType = listOf package;
+          in coercedTo fromType
+          (convertListToFunctionWithWarning "extraLuaPackages")
+          (functionTo fromType);
+        default = _: [ ];
+        defaultText = literalExpression "ps: [ ]";
+        example =
+          literalExpression "luaPackages: with luaPackages; [ luautf8 ]";
```

Comment:
the `luaPackages` may be confused with the top level one. You can use `ps` instead for `package set` or `lp` as `luaPackages` / `ls` (`lua set`)

Reply:
Better like this? `pyPkgs`/`luaPkgs`

Reply:
great, let's merge !